SpatialGDK/Source/SpatialGDK/Private/LoadBalancing/InterestManager.cpp

```diff
@@ -0,0 +1,615 @@
+		{
+			SwapEntries(ActiveEntities, Slot);
+			++ActiveEntities;
+		}
+	}
+}
+
+int32 FInterestManager::Allocate()
+{
+	Entities.AddDefaulted();
+	EntityPosition.AddDefaulted();
+	EntityPositionX.AddDefaulted();
+	EntityPositionY.AddDefaulted();
+	EntityFlags.AddDefaulted();
+	Visibility.AddDefaulted();
+	ActiveTimestamp.AddDefaulted();
+
+	++ActiveEntities;
+
+	int32 NumEntities = Entities.Num();
+	if (ActiveEntities != NumEntities)
+	{
+		MoveEntry(ActiveEntities - 1, NumEntities - 1);
+	}
+	return ActiveEntities - 1;
+}
+
+void FInterestManager::Remove(int32 Slot)
+{
+	if (Slot < ActiveEntities)
+	{
+		SwapEntries(ActiveEntities - 1, Slot);
+		Slot = ActiveEntities - 1;
+		--ActiveEntities;
+	}
+	if (Slot < Entities.Num() - 1)
+	{
+		SwapEntries(Slot, Entities.Num() - 1);
+	}
+	Entities.Pop(false);
+	EntityPosition.Pop(false);
+	EntityPositionX.Pop(false);
+	EntityPositionY.Pop(false);
+	EntityFlags.Pop(false);
+	Visibility.Pop(false);
+}
+
+void FInterestManager::MoveEntry(int32 SlotSrc, int32 SlotDest)
+{
+	Worker_EntityId EntitySrc = Entities[SlotSrc];
+
+	SlotMap[EntitySrc] = SlotDest;
+
+	Entities[SlotDest] = Entities[SlotSrc];
+	EntityPosition[SlotDest] = EntityPosition[SlotSrc];
+	EntityPositionX[SlotDest] = EntityPositionX[SlotSrc];
+	EntityPositionY[SlotDest] = EntityPositionY[SlotSrc];
+	EntityFlags[SlotDest] = EntityFlags[SlotSrc];
+	ActiveTimestamp[SlotDest] = ActiveTimestamp[SlotSrc];
+	Visibility[SlotDest] = Visibility[SlotSrc];
+}
+
+void FInterestManager::SwapEntries(int32 Slot1, int32 Slot2)
+{
+	Worker_EntityId Entity1 = Entities[Slot1];
+	Worker_EntityId Entity2 = Entities[Slot2];
+
+	SlotMap[Entity1] = Slot2;
+	SlotMap[Entity2] = Slot1;
+
+	Swap(Entities[Slot1], Entities[Slot2]);
+	Swap(EntityPosition[Slot1], EntityPosition[Slot2]);
+	Swap(EntityPositionX[Slot1], EntityPositionX[Slot2]);
+	Swap(EntityPositionY[Slot1], EntityPositionY[Slot2]);
+	Swap(EntityFlags[Slot1], EntityFlags[Slot2]);
+	Swap(ActiveTimestamp[Slot1], ActiveTimestamp[Slot2]);
+	Swap(Visibility[Slot1], Visibility[Slot2]);
+
+	if (Broadphase)
+	{
+		Broadphase->SwapEntries(Slot1, Slot2);
+	}
+}
+
+void FInterestManager::ComputeInterest(ISpatialOSWorker& Connection, const TArray<Worker_EntityId> Workers, const TArray<FBox2D>& Regions)
+{
+	SCOPE_CYCLE_COUNTER(STAT_InterestManagerComputation);
+
+	if (Entities.Num() == 0)
+	{
+		return;
+	}
+	if (Regions.Num() == 0)
+	{
+		return;
+	}
+
+	// Set an entity that has not moved since 5 secs as inactive. It will not be checked against the regions.
+	// This assumes that the bits allocation against workers remains constant, which is not valid anymore when changing/reassigning regions.
+	const double InactiveTime = 5.0;
+	const uint64 TimeToConsiderInactive = FPlatformTime::Cycles64() - (InactiveTime / FPlatformTime::GetSecondsPerCycle64());
+	for (int32 i = 0; i < (int32)ActiveEntities; ++i)
+	{
+		if (ActiveTimestamp[i] < TimeToConsiderInactive)
+		{
+			SwapEntries(i, ActiveEntities - 1);
```

Comment:
Ok so this pushes it out one element beyond the active entity bound in the slotmap